src/nvim/event/socket.c

```diff
@@ -40,17 +40,18 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
-    *host_end = '\0';
     uv_getaddrinfo_t request;
+    *host_end = '\0';
+    char *host = !strcmp(addr, "localhost") ? "127.0.0.1" : addr;
```

Comment:
If we use that `localhost` and `127.0.0.1` have the same length, the new array `host` is not needed:
```diff
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index d07b81bc2..bc80adc53 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -40,20 +40,24 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
-    uv_getaddrinfo_t request;
     *host_end = '\0';
-    char *host = !strcmp(addr, "localhost") ? "127.0.0.1" : addr;
+    uv_getaddrinfo_t request;
 
-    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, host, port,
+getaddrinfo:
+    ret = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
                                 &(struct addrinfo){
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
-    *host_end = ':';
-    if (retval != 0) {
-      ELOG("Host lookup failed: %s", host);
-      return retval;
+    if (ret != 0) {
+      ELOG("Host lookup failed: %s", addr);
+      if (strcmp(addr, "localhost") == 0) {
+        xstrlcpy(addr, "127.0.0.1", sizeof(addr));
+        goto getaddrinfo;
+      }
+      return ret;
     }
+    *host_end = ':';
     *port = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 

```